plugins/cpp/parser/src/clangastvisitor.h

```diff
@@ -1278,10 +1281,95 @@ class ClangASTVisitor : public clang::RecursiveASTVisitor<ClangASTVisitor>
     return getVisibility(decl_->getAccess());
   }
 
-  bool isWritten(const clang::VarDecl* vd_) const
+  bool isWritable(const clang::QualType& param) const
   {
-    // TODO
-    return true;
+    const clang::Type* typePtr = param.getTypePtr();
+
+    return
+      ((typePtr->isReferenceType() || typePtr->isPointerType()) &&
+       !typePtr->getPointeeType().isConstQualified()) ||
+      typePtr->isRValueReferenceType();
+  }
+
+  bool isWritten(const clang::Expr* expr_) const
+  {
+    while (expr_)
+    {
+      clang::ASTContext::DynTypedNodeList parents
+        = _astContext.getParents(*expr_);
+
+      const clang::ast_type_traits::DynTypedNode& parent = parents[0];
+
+      if (const clang::BinaryOperator* op = parent.get<clang::BinaryOperator>())
+      {
+        return
+          (op->isAssignmentOp() ||
+           op->isCompoundAssignmentOp() ||
+           op->isShiftAssignOp()) &&
+          op->getLHS() == expr_;
+      }
+      else if (const clang::UnaryOperator* op
+        = parent.get<clang::UnaryOperator>())
+      {
+        return op->isIncrementDecrementOp() && op->getSubExpr() == expr_;
+      }
+      else if (const clang::CallExpr* call = parent.get<clang::CallExpr>())
+      {
+        const clang::FunctionDecl* decl = call->getDirectCallee();
+
+        if (!decl)
+          return false;
+
+        const clang::CXXMethodDecl* method
+          = llvm::dyn_cast<clang::CXXMethodDecl>(decl);
+
+        if (method && !method->isConst() &&
+            call->getNumArgs() && call->getArg(0) == expr_)
+          return true;
+
+        unsigned x = method != nullptr;
```

Comment:
Add a short description for this variable (CallExpr first argument is `this` if its a CXXMethodDecl).